python-checks/src/main/java/org/sonar/python/checks/CognitiveComplexityFunctionCheck.java

```diff
@@ -0,0 +1,225 @@
+  offset = "5min",
+  effortToFixDescription = "per complexity point above the threshold")
+public class CognitiveComplexityFunctionCheck extends PythonCheck {
+
+  private static final String MESSAGE = "Refactor this method to reduce its Cognitive Complexity from %s to the %s allowed.";
+  public static final String CHECK_KEY = "S3776";
+  private static final int DEFAULT_THRESHOLD = 15;
+
+  private Set<AstNode> nestedFunctions = new HashSet<>();
+
+  private final CognitiveComplexityVisitor cognitiveComplexity = new CognitiveComplexityVisitor();
+  private AstWalker astWalker;
+
+  @RuleProperty(
+    key = "threshold",
+    description = "The maximum authorized complexity.",
+    defaultValue = "" + DEFAULT_THRESHOLD)
+  private int threshold = DEFAULT_THRESHOLD;
+
+  public void setThreshold(int threshold) {
+    this.threshold = threshold;
+  }
+
+  @Override
+  public void init() {
+    subscribeTo(PythonGrammar.FUNCDEF);
+    cognitiveComplexity.init();
+    astWalker = new AstWalker(cognitiveComplexity);
+  }
+
+  @Override
+  public void visitFile(@Nullable AstNode astNode) {
+    nestedFunctions.clear();
+  }
+
+  @Override
+  public void visitNode(AstNode astNode) {
+    if (nestedFunctions.contains(astNode)) {
+      return;
+    }
+
+    cognitiveComplexity.calculateComplexity(astNode);
+    int complexity = cognitiveComplexity.complexity;
+    if (complexity > threshold) {
+      String message = String.format(MESSAGE, complexity, threshold);
+      PreciseIssue issue = addIssue(astNode.getFirstChild(PythonGrammar.FUNCNAME), message)
+        .withCost(complexity - threshold);
+
+      cognitiveComplexity.secondaryLocations.forEach(issue::secondary);
+    }
+  }
+
+  private class CognitiveComplexityVisitor extends SquidAstVisitor<Grammar> {
+
+    private int complexity;
+    private AstNode functionNode;
+    private Set<IssueLocation> secondaryLocations = new HashSet<>();
+    private Map<AstNode, Integer> nestingLevelByNode = new HashMap<>();
+
+    @Override
+    public void init() {
+      subscribeTo(
+        PythonGrammar.IF_STMT,
+        PythonKeyword.ELIF,
+        PythonKeyword.ELSE,
+
+        PythonGrammar.WHILE_STMT,
+        PythonGrammar.FOR_STMT,
+
+        PythonGrammar.EXCEPT_CLAUSE,
+
+        PythonGrammar.FUNCDEF,
+
+        PythonGrammar.AND_TEST,
+        PythonGrammar.OR_TEST,
+
+        PythonGrammar.TEST
+      );
+    }
+
+    void calculateComplexity(AstNode funcDefNode) {
+      complexity = 0;
+      functionNode = funcDefNode;
+      nestingLevelByNode.clear();
+      secondaryLocations.clear();
+      nestingLevelByNode.put(functionNode.getFirstChild(PythonGrammar.SUITE), 0);
+      astWalker.walkAndVisit(funcDefNode);
+    }
+
+    @Override
+    public void visitNode(AstNode astNode) {
+      if (astNode.is(PythonGrammar.IF_STMT, PythonGrammar.WHILE_STMT, PythonGrammar.FOR_STMT)) {
+        incrementWithNestingAndNest(astNode.getFirstChild(), astNode.getFirstChild(PythonGrammar.SUITE));
```

Comment:
@vilchik-elena I'm not sure it would work but another approach would seem more natural to me. Instead of `nestingLevelByNode`, I would have a `nestingLevel` field which is incremented when entering a SUITE and decremented when leaving the SUITE (but not in `try: SUITE`).